pkg/gadgets/snapshot/socket/tracer/tracer.go

```diff
@@ -210,26 +214,29 @@ func RunCollector(pid uint32, podname, namespace, node string, proto socketcolle
 
 // ---
 
-type Tracer struct {
-	// visitedNamespaces is a map where the key is the netns inode number and
-	// the value is the pid of one of the containers that share that netns. Such
-	// pid is used by NetnsEnter. TODO: Improve NetnsEnter to also work with the
-	// netns directly.
-	visitedNamespaces map[uint64]uint32
-	protocols         string
-	eventHandler      func([]*socketcollectortypes.Event)
-}
-
-func (g *GadgetDesc) NewInstance() (gadgets.Gadget, error) {
+func NewTracer() (*Tracer, error) {
 	tracer := &Tracer{
 		visitedNamespaces: make(map[uint64]uint32),
+		protocols:         socketcollectortypes.ALL,
+		iters:             make(map[socketcollectortypes.Proto]*link.Iter),
 	}
 	return tracer, nil
 }
 
+func (g *GadgetDesc) NewInstance() (gadgets.Gadget, error) {
+	return NewTracer()
+}
+
 func (t *Tracer) Init(gadgetCtx gadgets.GadgetContext) error {
 	params := gadgetCtx.GadgetParams()
-	t.protocols = params.Get(ParamProto).AsString()
+	protocols := params.Get(ParamProto).AsString()
+
+	var ok bool
+	t.protocols, ok = socketcollectortypes.ProtocolsMap[protocols]
+	if !ok {
+		return fmt.Errorf("invalid proto %q", protocols)
+	}
```

Comment:
Yeah, good idea to use that - I think we had `Validate()` on the collections / params earlier, but removed it because we didn't use it (and we're now validating when setting the values).